### DISCUSSION TITLE
There is any way of not using rollup to bundle the code?

### DISCUSSION THREAD

**Discussion by gabrsar** [directus/directus#12625] [Archive]
It makes it impossible to debug, and handle stack traces as the files are minified and with dependencies code.

Also, many extensions that have the same dependencies bundle them many times in the js files making the overall code much bigger than it needs to be.

Wrapping it up, is any way of keeping my ts->js files split and using dependencies import in the 'normal' way?

Thank you.

**Comment by rijkvanzanten** [comment]
For the app extensions, the code needs to be bundled in a specific way so the app is able to register them in a known way. For API extensions, you can effectively do whatever you want, as long as it's CJS at the end of the day 🙂

**Reply by gabrsar** [reply]
@rijkvanzanten thank you a lot for your response.

I'm asking that because I had to make an integration using [sqs-consumer](https://www.npmjs.com/package/sqs-consumer) and I had some weird issues with the bundled code is generated and aws-sdk not being able to read properties. Also, is it conceptually ok to use background workers like this with directus in the same application?

The solution I found was to add the `sqs-consumer` lib as a dependency of the project itself and not on the extension, then injecting  a js file inside the directus folder that requires the lib and export it again, so I can "require" instead of importing this file on my extension and it is not bundled by rollup. But this is really messy and ugly.

Using the `default` way with rollup bundling it I had this error and found no way to fix this. Did you have any kind of experience with messed bundling before? Or it is just me? I tried to search but go no results that helped me.

Thank you a lot. Have a great time.


```
{ "level":40,"time":1649359378039,"pid":88,"hostname":"11585ea4721d","stack":"TypeError: Cannot read properties of undefined (reading 'memoizedProperty')\n    at Object.<anonymous> (/directus/extensions/hooks/extension-name-here/index.js:1:39920)\n    at Module._compile (node:internal/modules/cjs/loader:1103:14)\n    at Object.Module._extensions..js (node:internal/modules/cjs/loader:1155:10)\n    at Module.load (node:internal/modules/cjs/loader:981:32)\n    at Function.Module._load (node:internal/modules/cjs/loader:822:12)\n    at Module.require (node:internal/modules/cjs/loader:1005:19)\n    at require (node:internal/modules/cjs/helpers:102:18)\n    at ExtensionManager.registerHook (/directus/node_modules/directus/dist/extensions.js:250:30)\n    at ExtensionManager.registerHooks (/directus/node_modules/directus/dist/extensions.js:228:22)\n    at ExtensionManager.load (/directus/node_modules/directus/dist/extensions.js:134:14)","type":"Error","message":"Cannot read properties of undefined (reading 'memoizedProperty')"}
```

**Reply by rijkvanzanten** [reply]
Are you building an app or api extension? If API, you could simply ignore the extensions sdk / rollup setup 👍🏻

**Reply by gabrsar** [reply]
I'm using a hook extension to start an `sqs-consumer` with init('app.before') and keep it running in the background, so I think it's an app, not an API right? (still getting use to the names).

Thank you very much :)

**Reply by rijkvanzanten** [reply]
Nope! That's still an API extension 😄 The app extensions are any extensions that alter the admin app, like layouts and interfaces. Those need to be bundled in a specific way for our front-end bundle to be able to read and execute them properly. The API extensions (like hooks/endpoints) are imported using Node's regular CJS `require()`, so can be written in any way you like (as long as it's CJS compatible)

**Reply by Coladela** [reply]
@rijkvanzanten thank´s for your help !
Sorry I didn´t undertand what you mean with "simply ignore the extensions sdk / rollup setup" ? we should bundle a index file and just put on  extensions/[type] folder ?
Another questions is, following the purposed directus folder estructure and creating api extensions by cli, we have diferent npm packages, what is the best practice to share a code between different extensions ? 
Again, thank you for your help !

**Reply by rijkvanzanten** [reply]
> Sorry I didn´t undertand what you mean with "simply ignore the extensions sdk / rollup setup" ? we should bundle a index file and just put on extensions/[type] folder ?

If you're building an API extension, like a hook or an endpoint, you don't have to bundle anything. As long as you make sure the output is a valid node module export, you should be able to use whatever you want. For a lot of hooks / endpoints I use in debugging/examples, I just use "raw" javascript, for example:

```js
// /extensions/endpoints/my-endpoint/index.js

module.exports = function registerEndpoint(router) {
  router.get('/', (req, res) => res.send('Hello World'));
}
```

As long as the file is in the correct location, this will work.

> Another questions is, following the purposed directus folder estructure and creating api extensions by cli, we have diferent npm packages, what is the best practice to share a code between different extensions ?

That's a great question. Right now, each extension bundles to it's own output. @Nitwel, @nickrum and myself are actually currently investigating the idea of a "monorepo bundle" for extensions, where you can bundle (and therefore share logic/modules) multiple extensions into one distributable

**Reply by Coladela** [reply]
Ok I got you, thank you ! let me know about your investigation evolution ! Thank you !
